TimePicker: Add fiscal years and search to time picker

oscarkilhed opened the pull request:
### What this PR does / why we need it: ###

This adds fiscal year rounding to the time picker. It will now be possible to round to the start/end of a fiscal year or quarter using now/fy and now/fQ. When adding the fiscal time ranges the list got pretty long, so this also introduces a search function for the time picker.

Closes: https://github.com/grafana/grafana/issues/39052

oscarkilhed committed:
Add search to time picker

oscarkilhed committed:
implement fiscal datemath

oscarkilhed committed:
Fix for quarters

oscarkilhed committed:
Fix quarter tests and clean up quarter math

oscarkilhed committed:
Simplify year math

oscarkilhed committed:
Make collapsed time picker display correct phrase

dprokop committed:
Merge branch 'main' into oscark/time-picker-fiscal

dprokop commented on packages/grafana-data/src/datetime/datemath.ts:
```diff
@@ -102,11 +102,14 @@ export function parseDateMath(mathString: string, time: any, roundUp?: boolean):
   let i = 0;
   const len = strippedMathString.length;
 
+  const fy_start_month = 2;
```
Why 2 not 1? Is it common to have February as start of the fiscal year? Thought it's specific to Grafana

oscarkilhed replied:
This is just a dummy variable I'm using to until we've figured out what we want to do about the config. This should be passed into datemaths somehow.

dprokop replied:
I see, thanks!

dprokop commented on packages/grafana-data/src/datetime/datemath.ts:
```diff
@@ -159,3 +175,24 @@ export function parseDateMath(mathString: string, time: any, roundUp?: boolean):
   return dateTime;
 }
+
+export function roundToFiscal(FYStartMonth: number, dateTime: any, unit: string, roundUp: boolean | undefined) {
```
Nit: 
```suggestion
export function roundToFiscal(fyStartMonth: number, dateTime: any, unit: string, roundUp: boolean | undefined) {
```

oscarkilhed replied:
Ah, good catch

dprokop commented on public/app/core/components/PageActionBar/PageActionBar.tsx:
```diff
@@ -1,5 +1,5 @@
 import React, { PureComponent } from 'react';
-import { FilterInput } from '../FilterInput/FilterInput';
+import { FilterInput } from '@grafana/ui/src/components/FilterInput/FilterInput';
```
@mckn Do you think it would make sense to make the FilterInput a public component?

oscarkilhed replied:
I exposed the FilterInput under @grafana/ui and cleaned up the imports added by the VSCode refactor

dprokop commented on packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerContent.tsx:
```diff
@@ -179,26 +188,37 @@ export const TimePickerContentWithScreenSize: React.FC<PropsWithScreenSize> = (p
-          </CustomScrollbar>
+          <div className={styles.rightSide}>
+            <div className={styles.timeRangeFilter}>
+              <FilterInput width={0} value={searchTerm} onChange={setSearchQuery} placeholder={'Search quick ranges'} />
```
Noticed that when you press clear, the entire time picker disappears:

![Kapture 2021-09-16 at 10 22 00](https://user-images.githubusercontent.com/2376619/133577427-0a21027b-4d82-475b-95b0-82f378282883.gif)

dprokop replied:
One more nit: I would expect the search input to be autofocused when you open the time range picker so that you can start typing straight away :)

oscarkilhed replied:
Ah, I'll fix the clear and the focus, thanks

oscarkilhed replied:
https://github.com/grafana/grafana/pull/39073/commits/e7fa25eb068566500b465b76fa62d9ccf652c181
This should fix these issues

dprokop reviewed (commented):
UI looking good so far! Left few nits/comments

oscarkilhed committed:
Update packages/grafana-data/src/datetime/datemath.ts

Co-authored-by: Dominik Prokop <dominik.prokop@grafana.com>

oscarkilhed committed:
Change prop name FY -> fy

oscarkilhed committed:
Merge branch 'main' into oscark/time-picker-fiscal

oscarkilhed commented on packages/grafana-data/src/datetime/rangeutil.ts:
```diff
@@ -58,6 +60,10 @@ const rangeOptions: TimeOption[] = [
   { from: 'now-1y', to: 'now', display: 'Last 1 year' },
   { from: 'now-2y', to: 'now', display: 'Last 2 years' },
   { from: 'now-5y', to: 'now', display: 'Last 5 years' },
+  { from: 'now/fQ', to: 'now', display: 'This fiscal quarter so far' },
```
@marefr We want to add `f` as a way to round to fiscal quarters/years. However I've noticed that these strings gets sent to the backend. My first guess would be for alerting? I guess we need to implement the same parsing on the back end?

oscarkilhed committed:
Fix tests

oscarkilhed committed:
Remove unused import

oscarkilhed committed:
UX fixes on time picker

auto focus on search field
clear don't close the time picker
clear sets the focus on input field

oscarkilhed committed:
Expose FilterInput and change auto added imports

oscarkilhed committed:
add ui to change fiscal year setting

oscarkilhed committed:
Merge branch 'main' into oscark/time-picker-fiscal

oscarkilhed committed:
Adds fiscal year start date ui (very rough) and connects it to dashboard model

oscarkilhed committed:
Merge branch 'main' into oscark/time-picker-fiscal

oscarkilhed committed:
Hook up fiscal year setting to timesrv

oscarkilhed committed:
Add more months

oscarkilhed commented on packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerFooter.tsx:
```diff
@@ -43,42 +53,85 @@ export const TimePickerFooter: FC<Props> = (props) => {
+              <Field className={style.fiscalYearField} label={'Fiscal year start month'}>
+                <Select
+                  value={fiscalYearStartMonth}
+                  options={[
```
Perhaps this should be moved somewhere else? Maybe this exists elsewhere?

dprokop replied:
Maybe let's rename [`rangeOptions.ts`](https://github.com/grafana/grafana/blob/718d6fb25d61d73614e48982c01a7c0a52c42bd8/packages/grafana-ui/src/components/TimePicker/rangeOptions.ts) to `options.ts` and export these options from there?

oscarkilhed committed:
Update e2e test to match new time picker UI

dprokop commented on packages/grafana-ui/src/components/DateTimePickers/TimeRangePicker/TimePickerFooter.tsx:
```diff
@@ -43,42 +53,85 @@ export const TimePickerFooter: FC<Props> = (props) => {
+      <section aria-label="Time zone selection" className={style.container}>
+        <div className={style.timeZoneContainer}>
+          <div className={style.timeZone}>
+            <TimeZoneTitle title={info.name} />
+            <div className={style.spacer} />
+            <TimeZoneDescription info={info} />
```
I'm missing information about what's the currently selected fiscal year start month here. I know it's not in the designs, yet I think we should be slightly more verbose about what the selection is, similar to what we do with a time zone. Otherwise the fiscal year setting is super tricky to discover if you don't know where to look for it. What do you think @amy-super?

![image](https://user-images.githubusercontent.com/2376619/135077891-13134530-a062-4ec7-a237-dcfed0da6813.png)

dprokop commented on public/app/features/dashboard/state/DashboardModel.ts:
```diff
@@ -96,6 +96,7 @@ export class DashboardModel {
   panels: PanelModel[];
   panelInEdit?: PanelModel;
   panelInView?: PanelModel;
+  fiscalYearStartMonth?: number;
```
@hugohaggmark mind we are adding this new property to the DashboardModel.

hugohaggmark replied:
@dprokop thanks for heads up, does `0` mean January i.e. we're following default JavaScript month numbering confusion?

oscarkilhed replied:
Yes 0 is January. I was playing with the idea of having a month enum so we don't end up with negative or > 11 values. It would still be a 0 for Jan in the dashboard json though. Another option would be to hide it in month class with a string represenation for the json and then cast it when we need it for the moment.js integration. Feels a bit over engineered though

hugohaggmark replied:
Thanks for explaining that, maybe it's worth adding a brief comment about it?

dprokop reviewed (approved):
@oscarkilhed massive work on this one, looks great and works like a charm. I've left few nits regarding the UI/UX that you can talk about with @amy-super.

dprokop commented:
@oscarkilhed should we change this one from Draft to Ready for a review?

hugohaggmark commented on public/app/features/dashboard/state/DashboardModel.ts:
```diff
@@ -147,6 +148,7 @@ export class DashboardModel {
     this.refresh = data.refresh;
     this.snapshot = data.snapshot;
     this.schemaVersion = data.schemaVersion || 0;
+    this.fiscalYearStartMonth = data.fiscalYearStartMonth || 0;
```
super nit: Maybe use nullish coalescing operator instead? 
```suggestion
    this.fiscalYearStartMonth = data.fiscalYearStartMonth ?? 0;
```

oscarkilhed replied:
Ah, nice, I was just doing the same as the line above, coding on autopilot.

dprokop commented:
Ah, forgot about one more thing that @ryantxu and myself suggested. Can we get rid of the `Relative time ranges` and `Other quick ranges` headers now that we have search in the quick ranges picker?:

<img width="610" alt="Screen Shot 2021-09-28 at 13 35 10" src="https://user-images.githubusercontent.com/2376619/135079327-3745cf9b-374d-4c2a-b466-972981ace713.png">

oscarkilhed commented:
> Ah, forgot about one more thing that @ryantxu and myself suggested. Can we get rid of the `Relative time ranges` and `Other quick ranges` headers now that we have search in the quick ranges picker?:
> 
> <img alt="Screen Shot 2021-09-28 at 13 35 10" width="610" src="https://user-images.githubusercontent.com/2376619/135079327-3745cf9b-374d-4c2a-b466-972981ace713.png">

That would be nice as right now it has to be two list elements. With no headers it can be a single list.

oscarkilhed committed:
Make fiscal time ranges work for explore

amy-super commented:
> @dprokop
> 
> > Ah, forgot about one more thing that @ryantxu and myself suggested. Can we get rid of the `Relative time ranges` and `Other quick ranges` headers now that we have search in the quick ranges picker?:
> > <img alt="Screen Shot 2021-09-28 at 13 35 10" width="610" src="https://user-images.githubusercontent.com/2376619/135079327-3745cf9b-374d-4c2a-b466-972981ace713.png">
> 
> That would be nice as right now it has to be two list elements. With no headers it can be a single list.

That seems like a reasonable update - I kind of doubt those headers are helping much right now because they're quite difficult to see when you are quickly scanning through the list. And hopefully search becomes more of the go-to interaction.

oscarkilhed committed:
Make the ranges in the time picker a single list

oscarkilhed committed:
Fix tests

oscarkilhed committed:
Minor typo fix

oscarkilhed committed:
Add info tooltip for fiscal year

oscarkilhed committed:
Merge branch 'main' into oscark/time-picker-fiscal

torkelo commented:
not related to this PR but noticed that the FilterInput blocks of the placeholder text. I assume it's for the clear action, but that should not be needed when there is nothing in the input :) 

![Screenshot from 2021-09-29 14-26-35](https://user-images.githubusercontent.com/10999/135268175-f3b91edc-b669-4715-83b4-1adbc872db40.png)
t

torkelo commented:
I cannot use the keyboard to select the time ranges, I think this is pretty essential to make a search like this really useful and quick. Having a search input that support arrow up / down & enter to select seems like something that would be nice to have as a generic component that can be used for different list types :)

oscarkilhed commented:
> I cannot use the keyboard to select the time ranges, I think this is pretty essential to make a search like this really useful and quick. Having a search input that support arrow up / down & enter to select seems like something that would be nice to have as a generic component that can be used for different list types :)

I agree, I planned on adding that if we have time before the next release.

oscarkilhed committed:
Fix placeholder being blocked

oscarkilhed commented:
> not related to this PR but noticed that the FilterInput blocks of the placeholder text. I assume it's for the clear action, but that should not be needed when there is nothing in the input :)
> 
> ![Screenshot from 2021-09-29 14-26-35](https://user-images.githubusercontent.com/10999/135268175-f3b91edc-b669-4715-83b4-1adbc872db40.png) t

fixed in https://github.com/grafana/grafana/pull/39073/commits/e5b9d6c83aaddd9272332a980fce3e42bc237a85

torkelo commented:
Timepicker bug (not related to this), https://github.com/grafana/grafana/issues/39806

torkelo commented:
> fixed in e5b9d6c 

great!

oscarkilhed committed:
Merge branch 'main' into oscark/time-picker-fiscal
